var/spack/repos/builtin/packages/aspell/package.py

```diff
@@ -0,0 +1,83 @@
+##############################################################################
+# Copyright (c) 2013-2016, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the LICENSE file for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+import re
+from six.moves.urllib.parse import urlparse
+
+
+class Aspell(AutotoolsPackage):
+    """GNU Aspell is a Free and Open Source spell checker designed to 
+    eventually replace Ispell."""
+
+    homepage = "http://aspell.net/"
+    url      = "https://ftpmirror.gnu.org/aspell/aspell-0.60.6.1.tar.gz"
+
+    version('0.60.6.1', 'e66a9c9af6a60dc46134fdacf6ce97d7')
+
+    # additional dictionaries here: ftp://ftp.gnu.org/gnu/aspell/dict/0index.html
+    dicts = [
+        {'url': 'ftp://ftp.gnu.org/gnu/aspell/dict/en/aspell6-en-2017.01.22-0.tar.bz2',
+         'md5': "a6e002076574de9dc4915967032a1dab",
+        },
+        {'url': 'ftp://ftp.gnu.org/gnu/aspell/dict/es/aspell6-es-1.11-2.tar.bz2',
+         'md5': '8406336a89c64e47e96f4153d0af70c4',
+        },
+        {'url': 'ftp://ftp.gnu.org/gnu/aspell/dict/de/aspell6-de-20030222-1.tar.bz2',
+         'md5': '5950c5c8a36fc93d4d7616591bace6a6',
+        },
+    ]
```

Comment:
I can't make this work, my lack of understanding of python scoping behavior is showing...

I've touched up a few bits (finished the re.sub, yield name & package instead of p[4] and p[5], ...).

I moved the `dicts=...` definition inside of `def _list_resources()` so that it was visible in its `for` loop.

```python
    def list_resources():
        # additional dictionaries here: ftp://ftp.gnu.org/gnu/aspell/dict/0index.html
        dicts = [
            {'url': 'ftp://ftp.gnu.org/gnu/aspell/dict/en/aspell6-en-2017.01.22-0.tar.bz2',
             'md5': "a6e002076574de9dc4915967032a1dab",
            },
            {'url': 'ftp://ftp.gnu.org/gnu/aspell/dict/es/aspell6-es-1.11-2.tar.bz2',
             'md5': '8406336a89c64e47e96f4153d0af70c4',
            },
            {'url': 'ftp://ftp.gnu.org/gnu/aspell/dict/de/aspell6-de-20030222-1.tar.bz2',
             'md5': '5950c5c8a36fc93d4d7616591bace6a6',
            },
        ]
        for d in dicts:
            p = urlparse(d['url']).path.split('/')
            name = p[4]
            package = re.sub('\.tar.bz2', '', p[5])
            yield d['url'], d['md5'], name, package
```

Also can't use the name `url` in the loop that creates the variants and resources, it clashes mysteriously with the `url` definition on line 35.

If I use the @staticmethod decorator then the call in `for dict_url ...` loop fails with:

```
[...]
  File "/home/hartzelg/tmp/spack-aspell/var/spack/repos/builtin/packages/aspell/package.py", line 30, in <module>
    class Aspell(AutotoolsPackage):
  File "/home/hartzelg/tmp/spack-aspell/var/spack/repos/builtin/packages/aspell/package.py", line 59, in Aspell
    for dict_url,md5,name,package in list_resources():
TypeError: 'staticmethod' object is not callable
```

If I don't use the decorator then that bit works but the analogous bit installing the dictionary fails with:

```
[...]
==> Executing phase : 'install'
==> Error: NameError: global name 'list_resources' is not defined
/home/hartzelg/tmp/spack-aspell/var/spack/repos/builtin/packages/aspell/package.py:98, in install_dictionaries:
     95       @run_after('install')
     96       def install_dictionaries(self):
     97           prefix = self.prefix
  >> 98           for dict_url, md5, name, package in list_resources():
     99               if '+{0}'.format(name) in self.spec:
     100                  with working_dir(package):
[...]
```

Calling it as a method fails with:

```
==> Executing phase : 'build'
==> Executing phase : 'install'
==> Error: TypeError: list_resources() takes no arguments (1 given)
/home/hartzelg/tmp/spack-aspell/var/spack/repos/builtin/packages/aspell/package.py:98, in install_dictionaries:
     95       @run_after('install')
     96       def install_dictionaries(self):
     97           prefix = self.prefix
  >> 98           for dict_url, md5, name, package in self.list_resources():
     99               if '+{0}'.format(name) in self.spec:
```